src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/Connection/PreloginSubHandlers/PreloginPacketHandler.cs

```diff
@@ -0,0 +1,412 @@
+                        // DO NOTHING FOR TRACEID
+                        offset += PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES;
+                        break;
+
+                    case (int)PreLoginOptions.FEDAUTHREQUIRED:
+                        payloadOffset = payload[offset++] << 8 | payload[offset++];
+                        offset += 2; // Skip the payload length
+
+                        // Only 0x00 and 0x01 are accepted values from the server.
+                        if (payload[payloadOffset] != 0x00 && payload[payloadOffset] != 0x01)
+                        {
+                            SqlClientEventSource.Log.TryTraceEvent("<sc.{0}|ERR> {1}, " +
+                                "Server sent an unexpected value for FedAuthRequired PreLogin Option. Value was {2}.", nameof(ReadPreLoginresponse), ObjectID, (int)payload[payloadOffset]);
+                            throw SQL.ParsingErrorValue(ParsingErrorState.FedAuthRequiredPreLoginResponseInvalidValue, (int)payload[payloadOffset]);
+                        }
+
+                        // We must NOT use the response for the FEDAUTHREQUIRED PreLogin option, if the connection string option
+                        // was not using the new Authentication keyword or in other words, if Authentication=NotSpecified
+                        // Or AccessToken is not null, mean token based authentication is used.
+                        if ((context.ConnectionContext.ConnectionString != null
+                            && context.ConnectionContext.ConnectionString.Authentication != SqlAuthenticationMethod.NotSpecified)
+                            || context.ConnectionContext.AccessTokenInBytes != null || context.ConnectionContext.AccessTokenCallback != null)
+                        {
+                            context.ConnectionContext.FedAuthRequired = payload[payloadOffset] == 0x01;
+                        }
+                        break;
+
+                    default:
+                        Debug.Fail("UNKNOWN option in ConsumePreLoginHandshake, option:" + option);
+                        // DO NOTHING FOR THESE UNKNOWN OPTIONS
+                        offset += PAYLOAD_OFFSET_AND_LENGTH_SIZE_IN_BYTES;
+                        break;
+                }
+
+                if (offset < payload.Length)
+                {
+                    option = payload[offset++];
+                }
+                else
+                {
+                    break;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Constructs the Prelogin packet and sends it to the server.
+        /// </summary>
+        /// <param name="context">The Prelogin handler context</param>
+        /// <param name="isAsync"></param>
+        /// <param name="ct"></param>
+        /// <returns></returns>
+        private static async Task CreatePreLoginAndSend(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
```

Comment:
```suggestion
        private static async ValueTask CreatePreLoginAndSend(PreLoginHandlerContext context, bool isAsync, CancellationToken ct)
```

Reply:
@edwardneal, although ValueTask optimizes the perf and avoids unnecessary allocation, but it will add some more complexity that it may introduce some potential bugs. Can network latency have any effect on ValueTask vs Task?

Reply:
In the exclusively async case, we'd almost always use Task rather than ValueTask; this network traffic is almost always going to be non-blocking by definition, regardless of latency. I'm specifying ValueTask so that we don't need to allocate Tasks if isAsync is false and we run synchronously.